synapse/util/iterutils.py

```diff
@@ -71,6 +73,31 @@ def chunk_seq(iseq: S, maxlen: int) -> Iterator[S]:
     return (iseq[i : i + maxlen] for i in range(0, len(iseq), maxlen))
 
 
+def partition(
+    iterable: Iterable[T], predicate: Callable[[T], bool]
+) -> Tuple[List[T], List[T]]:
+    """
+    Separate a given iterable into two lists based on the result of a predicate function.
+
+    Args:
+        iterable: the iterable to partition (separate)
+        predicate: a function that takes an item from the iterable and returns a boolean
+
+    Returns:
+        A tuple of two lists, the first containing all items for which the predicate
+        returned True, the second containing all items for which the predicate returned
+        False
+    """
+    true_results = []
+    false_results = []
+    for item in iterable:
+        if predicate(item):
+            true_results.append(item)
+        else:
+            false_results.append(item)
+    return true_results, false_results
```

Comment:
[@clokep was wishing](https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$3ygksoSclc6ZlWNCKqysasK3oU_lrLkdeOL6nveDliA?via=matrix.org&via=element.io&via=pixie.town) generally for a `partition(...)` function being built-in to `itertools` and being useful in many scenarios. I decided to add it here since we already have a dedicated `synapse/util/iterutils.py` file. 

`partition(...)` not only helps with double iteration when separating a list naively with list comprehensions but more importantly for clarity on your intention.

I used a more simple version of `partition(...)`  from https://stackoverflow.com/a/4578605/796832 vs the [`itertools` recipe](https://docs.python.org/dev/library/itertools.html#itertools-recipes) because it seemed more straightforward. But if people have preference for the more clever `itertools` version, I have no problem updating to use that instead.

Reply:
This version is definitely clearer! I can't say the other version has any strong benefits?